lib/postgres_v3_experimental.dart

```diff
@@ -0,0 +1,204 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+
+abstract class PgPool implements PgSession, PgSessionExecutor {
+  static Future<PgPool> open(
+    List<PgEndpoint> endpoints, {
+    PgSessionSettings? sessionSettings,
+    PgPoolSettings? poolSettings,
+  }) =>
+      throw UnimplementedError();
+
+  Future<R> withConnection<R>(
+    Future<R> Function(PgConnection connection) fn, {
+    PgSessionSettings? sessionSettings,
+  });
+}
+
+abstract class PgSession {
+  // uses extended query protocol
+  Future<PgStatement> prepare(
+    String sql, {
+    Object? /* String */ substitution,
+    Object? /* List<PgDataType> | Map<String, PgDataType> */ types,
+    Duration? timeout,
+  });
+
+  Future<PgResult> execute(
+    String sql, {
+    Object? /* String */ substitution,
```

Comment:
What does `substitution` do? I guess it's related to `types` and `parameters`, but in which way exactly? Don't these two provide all the information we need?

Reply:
I thought that to get flexibility in the query substitution, we could have both `?` or `:name` or `@name` as substitution pattern. e.g. `substitution: ':'` would enable `:name` replacement patterns, keeping the `@` out of harms way...

Reply:
I think it's a good idea to allow multiple formats, but at least to me it was confusing that they are all parameters on the same method, especially since they appear to be somewhat incompatible with each other. Maybe we could have something like

```dart
class PgQueryDescription {
  factory PgQueryDescription.direct(String sql, {List<PgDataType>? types}); // direct postgres api
  factory PgQueryDescription.map(String sql, {String substitution = '@'}); // local post-processing
}
```

And then `prepare` would take a `PGQueryDescription` instead of `sql` + `types` + `substitution`? To make the default case easier, we could also have it take an `Object /* String | PgQueryDescription */`, defaulting to `PgQueryDescription.map` if given a string.

Reply:
Good points here. I tend to agree (as the alternative would be `prepareDirect`, `prepareMap`, ...), although I'd probably go with a much shorter name, like `PgSql`, as this would be used frequently.

Another angle for this would be to configure the substitution only at the connection/session level.